interceptor/tracing_interceptor.go

```diff
@@ -847,14 +930,54 @@ func (t *tracingInterceptor) startSpan(
 	}
 
 	// Put span in header if wanted
-	if options.ToHeader && header != nil {
-		if err := t.writeSpanToHeader(span, header); err != nil {
+	if options.ToHeader {
+		if err := headerWriter(span); err != nil {
 			return nil, err
 		}
 	}
 	return span, nil
 }
 
+func (t *tracingInterceptor) headerReader(ctx context.Context) func() (TracerSpanRef, error) {
```

Comment:
Pedantic since it's not exported, but usually Go would prefer an interface to passing multiple functions around, e.g.

```go
type spanMarshaler interface {
	readSpan(*tracingInterceptor) (TracerSpanRef, error)
	writeSpan(*tracingInterceptor, TracerSpan) error
}

type payloadSpanMarshaler map[string]*commonpb.Payload
func (p payloadSpanMarshaler) readSpan(t *tracingInterceptor) (TracerSpanRef, error) {
	return t.readSpanFromHeader(p)
}
func (p payloadSpanMarshaler) writeSpan(t *tracingInterceptor, span TracerSpan) error {
	return t.writeSpanToHeader(span, p)
}

type stringSpanMarshaler map[string]string

// ...etc
```

May be able to just inline `readSpanFromHeader` and `writeSpanToHeader` too, but pedantic